In [1]:
import numpy as np
import pandas as pd

In [33]:
gameplay_df = pd.read_csv(r"C:\Users\ADITYA SINGH\OneDrive\Desktop\DataMining\Analytics Position Case Study\Analytics Position Case Study - User Gameplay data.csv",skiprows=2,header = 1 )

In [35]:
gameplay_df.head()

User ID  Games Played         Datetime
0      851             1  01-10-2022 0:00
1      717             1  01-10-2022 0:00
2      456             1  01-10-2022 0:00
3      424             1  01-10-2022 0:00
4      845             1  01-10-2022 0:00

In [65]:
gameplay_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 355266 entries, 0 to 355265
Data columns (total 3 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   User ID       355266 non-null  int64 
 1   Games Played  355266 non-null  int64 
 2   Datetime      355266 non-null  object
dtypes: int64(2), object(1)
memory usage: 8.1+ MB


In [69]:
gameplay_df['Datetime'] = pd.to_datetime(gameplay_df['Datetime'],format='%d-%m-%Y %H:%M')

In [71]:
gameplay_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 355266 entries, 0 to 355265
Data columns (total 3 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   User ID       355266 non-null  int64         
 1   Games Played  355266 non-null  int64         
 2   Datetime      355266 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(2)
memory usage: 8.1 MB


In [47]:
deposit_df = pd.read_csv(r"C:\Users\ADITYA SINGH\OneDrive\Desktop\DataMining\Analytics Position Case Study\Analytics Position Case Study - Deposit Data.csv",skiprows=2,header=1)

In [49]:
deposit_df.head()

User Id         Datetime  Amount
0      357  01-10-2022 0:03    2000
1      776  01-10-2022 0:03    2500
2      492  01-10-2022 0:06    5000
3      803  01-10-2022 0:07    5000
4      875  01-10-2022 0:09    1500

In [73]:
deposit_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17438 entries, 0 to 17437
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   User Id   17438 non-null  int64 
 1   Datetime  17438 non-null  object
 2   Amount    17438 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 408.8+ KB


In [75]:
deposit_df['Datetime'] = pd.to_datetime(deposit_df['Datetime'],format='%d-%m-%Y %H:%M')

In [77]:
deposit_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17438 entries, 0 to 17437
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   User Id   17438 non-null  int64         
 1   Datetime  17438 non-null  datetime64[ns]
 2   Amount    17438 non-null  int64         
dtypes: datetime64[ns](1), int64(2)
memory usage: 408.8 KB


In [53]:
withdrawal_df = pd.read_csv(r"C:\Users\ADITYA SINGH\OneDrive\Desktop\DataMining\Analytics Position Case Study\Analytics Position Case Study - Withdrawal Data.csv",skiprows=2,header=1)

In [55]:
withdrawal_df.head()

User Id         Datetime  Amount
0      190  01-10-2022 0:03    5872
1      159  01-10-2022 0:16    9540
2      164  01-10-2022 0:24     815
3      946  01-10-2022 0:29   23000
4      763  01-10-2022 0:40    9473

In [79]:
withdrawal_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3566 entries, 0 to 3565
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   User Id   3566 non-null   int64 
 1   Datetime  3566 non-null   object
 2   Amount    3566 non-null   int64 
dtypes: int64(2), object(1)
memory usage: 83.7+ KB


In [81]:
withdrawal_df['Datetime'] = pd.to_datetime(withdrawal_df['Datetime'],format='%d-%m-%Y %H:%M')

In [83]:
withdrawal_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3566 entries, 0 to 3565
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   User Id   3566 non-null   int64         
 1   Datetime  3566 non-null   datetime64[ns]
 2   Amount    3566 non-null   int64         
dtypes: datetime64[ns](1), int64(2)
memory usage: 83.7 KB


In [59]:
gameplay_df.shape

(355266, 3)

In [61]:
deposit_df.shape

(17438, 3)

In [63]:
withdrawal_df.shape

(3566, 3)

In [105]:
def get_slot(dt):
    hour = dt.hour
    if 0<= hour <12:
        return "S1"
    else:
        return "S2"

In [107]:
gameplay_df["Slot"] = gameplay_df['Datetime'].apply(get_slot)

In [111]:
deposit_df["Slot"] = deposit_df['Datetime'].apply(get_slot)

In [115]:
withdrawal_df["Slot"] = withdrawal_df['Datetime'].apply(get_slot)

In [135]:
gameplay_df['User ID'].value_counts()

User ID
765    24096
618     7084
247     6642
663     6401
449     6300
       ...  
25         1
661        1
543        1
848        1
969        1
Name: count, Length: 1000, dtype: int64

In [143]:
y = deposit_df['User Id'].unique()

In [147]:
np.isin(y,x).all()   #So every player from deposit table has played game

True

But as number of unique players in gameplay table is more than number of unique players in deposit table. So we will use gameplay for getting all player's user_id

In [152]:
UserId = gameplay_df['User ID'].unique()    #col1

In [156]:
len(UserId)

1000

 ## 	Part A - Calculating loyalty points

In [196]:
#function for filtering dataframe according to date and slot
def filtering_data(df,date_,S):
    date_ = pd.to_datetime(date_).date()
    res = df[(df['Datetime'].dt.date == date_) & (df['Slot'] == S)]
    return res

In [323]:
#function for calculating Loyalty points 
def Loyalty_points(A,B,C):
    Games_Played = A.groupby('User ID')['Games Played'].sum()                 #sum of games played by each player
    Deposit_Amount = B.groupby('User Id')['Amount'].sum()                     #sum of deposited amount by each player
    Withdrawal_Amount = C.groupby('User Id')['Amount'].sum()                  #sum of withdrawal amount by each player
    number_deposits = B.groupby('User Id')['Amount'].count()                  #number of deposits per player
    number_withdrawals = C.groupby('User Id')['Amount'].count()               #number of withdrawals per player
    df = pd.concat([Games_Played,Deposit_Amount,Withdrawal_Amount,number_deposits,number_withdrawals], axis = 1)
    df.columns = ['Games Played', 'Deposit Amount', 'Withdrawal Amount', 
                  'Number of Deposits', 'Number of Withdrawals']
    #Converting data types to numeric as i was getting some errors dur to it.
    
    df['Deposit Amount'] = pd.to_numeric(df['Deposit Amount'], errors='coerce')
    df['Withdrawal Amount'] = pd.to_numeric(df['Withdrawal Amount'], errors='coerce')
    df['Games Played'] = pd.to_numeric(df['Games Played'], errors='coerce')
    df['Number of Deposits'] = pd.to_numeric(df['Number of Deposits'], errors='coerce')
    df['Number of Withdrawals'] = pd.to_numeric(df['Number of Withdrawals'], errors='coerce')
    
    df.fillna(0,inplace=True)
    Loyalty_Points = (0.01 * df['Deposit Amount']) + (0.005 * df['Withdrawal Amount']) + (0.001 * (df['Number of Deposits'] - df['Number of Withdrawals']).clip(lower=0)) + (0.2 * df['Games Played'])
    df['Loyalty Points'] = Loyalty_Points
    return df

2nd October Slot S1

In [326]:
A1 = filtering_data(gameplay_df,'2022-10-02','S1')

In [328]:
B1 = filtering_data(deposit_df,'2022-10-02','S1')

In [330]:
C1 = filtering_data(withdrawal_df,'2022-10-02','S1')

In [332]:
Loyalty_points(A1,B1,C1)

Games Played  Deposit Amount  Withdrawal Amount  Number of Deposits  \
2             2.0             0.0                0.0                 0.0   
5             9.0           600.0                0.0                 1.0   
6             2.0          4000.0                0.0                 1.0   
7             5.0             0.0                0.0                 0.0   
8             6.0             0.0                0.0                 0.0   
..            ...             ...                ...                 ...   
435           0.0             0.0             1600.0                 0.0   
560           0.0             0.0             7797.0                 0.0   
634           0.0             0.0           295671.0                 0.0   
936           0.0             0.0             6857.0                 0.0   
944           0.0             0.0              500.0                 0.0   

     Number of Withdrawals  Loyalty Points  
2                      0.0           0.400  
5                      0.0           7.801  
6                      0.0          40.401  
7                      0.0           1.000  
8                      0.0           1.200  
..                     ...             ...  
435                    1.0           8.000  
560                    1.0          38.985  
634                    1.0        1478.355  
936                    1.0          34.285  
944                    1.0           2.500  

[636 rows x 6 columns]

16th October Slot S2

In [336]:
A2 = filtering_data(gameplay_df,'2022-10-16','S2')

In [338]:
B2 = filtering_data(deposit_df,'2022-10-16','S2')

In [340]:
C2 = filtering_data(withdrawal_df,'2022-10-16','S2')

In [342]:
Loyalty_points(A2,B2,C2)

Games Played  Deposit Amount  Withdrawal Amount  Number of Deposits  \
2             2.0             0.0                0.0                 0.0   
5             9.0          1000.0                0.0                 2.0   
6             1.0             0.0                0.0                 0.0   
8             5.0             0.0                0.0                 0.0   
9            51.0          2000.0                0.0                 1.0   
..            ...             ...                ...                 ...   
468           0.0             0.0             7500.0                 0.0   
634           0.0             0.0           298311.0                 0.0   
762           0.0             0.0             9988.0                 0.0   
786           0.0             0.0           120000.0                 0.0   
947           0.0             0.0            10000.0                 0.0   

     Number of Withdrawals  Loyalty Points  
2                      0.0           0.400  
5                      0.0          11.802  
6                      0.0           0.200  
8                      0.0           1.000  
9                      0.0          30.201  
..                     ...             ...  
468                    1.0          37.500  
634                    1.0        1491.555  
762                    1.0          49.940  
786                    1.0         600.000  
947                    1.0          50.000  

[594 rows x 6 columns]

 18th October Slot S1

In [344]:
A3 = filtering_data(gameplay_df,'2022-10-18','S1')

In [346]:
B3 = filtering_data(deposit_df,'2022-10-18','S1')

In [348]:
C3 = filtering_data(withdrawal_df,'2022-10-18','S1')

In [350]:
Loyalty_points(A3,B3,C3)

Games Played  Deposit Amount  Withdrawal Amount  Number of Deposits  \
2             2.0             0.0                0.0                 0.0   
3             2.0             0.0                0.0                 0.0   
5             8.0             0.0                0.0                 0.0   
7             3.0             0.0                0.0                 0.0   
8             9.0             0.0                0.0                 0.0   
..            ...             ...                ...                 ...   
441           0.0             0.0            10000.0                 0.0   
553           0.0             0.0             8500.0                 0.0   
634           0.0             0.0           544620.0                 0.0   
691           0.0             0.0              510.0                 0.0   
817           0.0             0.0             4000.0                 0.0   

     Number of Withdrawals  Loyalty Points  
2                      0.0            0.40  
3                      0.0            0.40  
5                      0.0            1.60  
7                      0.0            0.60  
8                      0.0            1.80  
..                     ...             ...  
441                    1.0           50.00  
553                    1.0           42.50  
634                    2.0         2723.10  
691                    1.0            2.55  
817                    1.0           20.00  

[622 rows x 6 columns]

26th October Slot S2

In [352]:
A4 = filtering_data(gameplay_df,'2022-10-26','S2')

In [354]:
B4 = filtering_data(deposit_df,'2022-10-26','S2')

In [356]:
C4 = filtering_data(withdrawal_df,'2022-10-26','S2')

In [358]:
Loyalty_points(A4,B4,C4)

Games Played  Deposit Amount  Withdrawal Amount  Number of Deposits  \
5             4.0          1000.0                0.0                 1.0   
6             1.0             0.0                0.0                 0.0   
7             1.0             0.0             6617.0                 0.0   
8             7.0             0.0                0.0                 0.0   
9            44.0          1000.0            20076.0                 1.0   
..            ...             ...                ...                 ...   
158           0.0             0.0             5000.0                 0.0   
245           0.0             0.0            70000.0                 0.0   
274           0.0             0.0            11000.0                 0.0   
402           0.0             0.0            12280.0                 0.0   
949           0.0             0.0            40000.0                 0.0   

     Number of Withdrawals  Loyalty Points  
5                      0.0          10.801  
6                      0.0           0.200  
7                      1.0          33.285  
8                      0.0           1.400  
9                      1.0         119.180  
..                     ...             ...  
158                    1.0          25.000  
245                    1.0         350.000  
274                    1.0          55.000  
402                    1.0          61.400  
949                    1.0         200.000  

[628 rows x 6 columns]

2. Calculate overall loyalty points earned and rank players on the basis of loyalty points in the month of October. 
     In case of tie, number of games played should be taken as the next criteria for ranking.

In [361]:
def filtering_data2(df,month_):
    res = df[(df['Datetime'].dt.month == month_)]
    return res

In [393]:
def Loyalty_points2(A,B,C):
    Games_Played = A.groupby('User ID')['Games Played'].sum()                 #sum of games played by each player
    Deposit_Amount = B.groupby('User Id')['Amount'].sum()                     #sum of deposited amount by each player
    Withdrawal_Amount = C.groupby('User Id')['Amount'].sum()                  #sum of withdrawal amount by each player
    number_deposits = B.groupby('User Id')['Amount'].count()                  #number of deposits per player
    number_withdrawals = C.groupby('User Id')['Amount'].count()               #number of withdrawals per player
    df = pd.concat([Games_Played,Deposit_Amount,Withdrawal_Amount,number_deposits,number_withdrawals], axis = 1)
    df.columns = ['Games Played', 'Deposit Amount', 'Withdrawal Amount', 
                  'Number of Deposits', 'Number of Withdrawals']
    
    df['Deposit Amount'] = pd.to_numeric(df['Deposit Amount'], errors='coerce')
    df['Withdrawal Amount'] = pd.to_numeric(df['Withdrawal Amount'], errors='coerce')
    df['Games Played'] = pd.to_numeric(df['Games Played'], errors='coerce')
    df['Number of Deposits'] = pd.to_numeric(df['Number of Deposits'], errors='coerce')
    df['Number of Withdrawals'] = pd.to_numeric(df['Number of Withdrawals'], errors='coerce')
    
    df.fillna(0,inplace=True)
    Loyalty_Points = (0.01 * df['Deposit Amount']) + (0.005 * df['Withdrawal Amount']) + (0.001 * (df['Number of Deposits'] - df['Number of Withdrawals']).clip(lower=0)) + (0.2 * df['Games Played'])
    df['Loyalty Points'] = Loyalty_Points
    #till here its same as previous one
    df = df.sort_values(by=['Loyalty Points','Games Played'] , ascending = [False,False])
    df['Rank'] = df['Loyalty Points'].rank(method='dense', ascending=False)
    #df['Rank'] = range(1, len(df) + 1)
    return df

In [395]:
A5 = filtering_data2(gameplay_df,10)

In [397]:
B5 = filtering_data2(deposit_df,10)

In [399]:
C5 = filtering_data2(withdrawal_df,10)

In [465]:
Rankings_of_month = Loyalty_points2(A5,B5,C5)
Rankings_of_month

Games Played  Deposit Amount  Withdrawal Amount  Number of Deposits  \
634            24        515000.0         15737705.0                 8.0   
99             10       1164800.0          2403141.0                47.0   
672            10       2158700.0           233750.0                35.0   
212             1       1924981.0           589850.0                26.0   
740             2       1738490.0           365288.0                91.0   
..            ...             ...                ...                 ...   
643             2             0.0                0.0                 0.0   
857             2             0.0                0.0                 0.0   
858             2             0.0                0.0                 0.0   
993             2             0.0                0.0                 0.0   
38              1             0.0                0.0                 0.0   

     Number of Withdrawals  Loyalty Points   Rank  
634                   67.0       83843.325    1.0  
99                    15.0       23665.737    2.0  
672                    5.0       22757.780    3.0  
212                    4.0       22199.282    4.0  
740                    7.0       19211.824    5.0  
..                     ...             ...    ...  
643                    0.0           0.400  963.0  
857                    0.0           0.400  963.0  
858                    0.0           0.400  963.0  
993                    0.0           0.400  963.0  
38                     0.0           0.200  964.0  

[1000 rows x 7 columns]

3. What is the average deposit amount?

In [406]:
avg = (deposit_df['Amount'].sum()) / (len(deposit_df))

In [408]:
avg

5492.185399701801

4. What is the average deposit amount per user in a month?

In [419]:
X = deposit_df['Datetime'].dt.month             #checking whether dataframe has more than one month
X.nunique()

1

In [433]:
def avg_deposit(data):
    res = data.groupby('User Id')['Amount'].mean()
    return pd.DataFrame(res)

In [435]:
avg_deposit(deposit_df)

Amount
User Id              
1         5000.000000
2        28350.000000
3        10000.000000
4         1750.000000
5         1105.970149
...               ...
994       1400.000000
995       6900.000000
996       3109.677419
998       1480.000000
999       3000.000000

[915 rows x 1 columns]

5. What is the average number of games played per user?

In [446]:
def avg_gamePlayed(data):
    res = data.groupby('User ID')['Games Played'].mean()
    return pd.DataFrame(res)

In [460]:
X = avg_gamePlayed(gameplay_df)
X.value_counts()

Games Played
1.000000        999
1.000042          1
Name: count, dtype: int64

## Part B - How much bonus should be allocated to leaderboard players?

After calculating the loyalty points for the whole month find out which 50 players are at the top of the leaderboard. The company has allocated a pool of Rs 50000 to be given away as bonus money to the loyal players.

Now the company needs to determine how much bonus money should be given to the players.

Should they base it on the amount of loyalty points? Should it be based on number of games? Or something else?

That’s for you to figure out.

Suggest a suitable way to divide the allocated money keeping in mind the following points:
1. Only top 50 ranked players are awarded bonus

In [469]:
Top_50Ranked_players = Rankings_of_month[Rankings_of_month['Rank'] <= 50]
Top_50Ranked_players

Games Played  Deposit Amount  Withdrawal Amount  Number of Deposits  \
634            24        515000.0         15737705.0                 8.0   
99             10       1164800.0          2403141.0                47.0   
672            10       2158700.0           233750.0                35.0   
212             1       1924981.0           589850.0                26.0   
740             2       1738490.0           365288.0                91.0   
566           183       1819175.0           185071.0                53.0   
714             6       1676300.0                0.0                34.0   
421          1557        878600.0          1269809.0                99.0   
369            37        650000.0          1586208.0                13.0   
30             13       1329000.0           152145.0                51.0   
587           734        671700.0          1355000.0                89.0   
222            10       1285000.0            99358.0                16.0   
352           313       1083034.0           429518.0               128.0   
365          3667        311000.0          1802335.0                10.0   
920           932        367700.0          1734480.0                41.0   
162             2         68320.0          2360000.0                 7.0   
415            26        350000.0          1759843.0                 6.0   
569            38       1227780.0                0.0                23.0   
786             7        661500.0          1096160.0                27.0   
2              97        567000.0          1270215.0                20.0   
238           899        923400.0           445000.0                45.0   
992          2483        804978.0           616278.0                18.0   
28             37       1061500.0                0.0                49.0   
538            29       1025000.0             8558.0                27.0   
208           507        644934.0           657928.0                35.0   
989          3471        240000.0          1339000.0                12.0   
978            74        794358.0           288114.0                69.0   
915             8        780028.0           307609.0                84.0   
678            12        379000.0          1100000.0                22.0   
78            602        159000.0          1484900.0                32.0   
909             1        679745.0           454060.0               125.0   
182          2446        795000.0            80339.0                16.0   
93            487        616000.0           511474.0                76.0   
200             4        539998.0           593000.0                 8.0   
259           126        824020.0                0.0               115.0   
306             4        826000.0                0.0                57.0   
344             3        311900.0          1024000.0                18.0   
601            33        692144.0           240000.0                23.0   
515           413        580000.0           400000.0                14.0   
294             5        607700.0           311187.0                41.0   
681             8        689000.0           132000.0                65.0   
950             4        732000.0            35500.0                99.0   
675           639        662500.0            70000.0                50.0   
663          6401        425000.0           290200.0                52.0   
438             1        400000.0           594851.0                 2.0   
619             2        697000.0                0.0                13.0   
245             3        209450.0           912200.0                25.0   
547          1248        493200.0           245863.0                41.0   
612          1543        554800.0           109081.0                53.0   
949             1        409100.0           434171.0                64.0   

     Number of Withdrawals  Loyalty Points  Rank  
634                   67.0       83843.325   1.0  
99                    15.0

ANS --> 
- According to me, the primary method for allocation should be `LOYALTY POINT` as it takes into account `Deposit amount`, `number of deposits`, `withdrawable amount` and `no of games played` as well.
(On the other hand if we select `game played` as a sole criteria for allocation , we may miss the chance of attracting players who deposit more)
        -then Use `games played` as a secondary criteria for allocation.
        lets take an example
        85% of money should be distributed on the basis of loyalty points (as it will welcome the profit) (X Rs) and 15%  should be distributed on the basis of games played (it will improve engagement time ) (Y Rs)                                     
         -->Loyalty Bonus for Player=( Player’s Loyalty Points / Total Loyalty Points of Top 50) * X
         -->Games Bonus for Player=( Player’s Games Played / Total Games Played of Top 50 ) * Y
                                        

## PART C -> Would you say that loyalty formula is fair or unfair?
Can you suggest any way to make it more robust?

Loyalty Point = (0.01 * deposit) + (0.005 * Withdrawal amount) + (0.001 * (maximum of (#deposit - #withdrawal) or 0)) + (0.2 * Number of games played)
I believe that `(0.005 * Withdrawal amount)` portion should be minimized or excluded and it's weightage should be added to `game played`. As it would encourage engagement.